sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1031,14 +1072,24 @@ Exceptions:
 
 * Throws synchronously with error code `invalid` if a queue is recording
   commands to the graph.
+
 * Throws synchronously with error code `invalid` if the graph wasn't created with
   the `property::graph::assume_buffer_outlives_graph` property and this command
   uses a buffer. See the
   <<assume-buffer-outlives-graph-property, Assume-Buffer-Outlives-Graph>>
   property for more information.
+
 * Throws with error code `invalid` if the type of the command-group is not a
   kernel execution and a `dynamic_parameter` was registered inside `cgf`.
 
+* Throws synchronously with error code `invalid` if an `event` dependency is
```

Comment:
Maybe I misunderstood something, but isn't this inconsistent with the line 283:

```
However, for an event passed to `handler::depends_on()` to
create an edge, it must be an event returned from a queue submission captured by
the same graph. Passing events from other sources (other graph submissions,
egular SYCL submissions) will not create edges in the graph, but will create
runtime dependencies for a graph node on those other events.
```

Doesn't this imply that passing an event from a recorded graph should be acceptable ? Is this a difference between:

event: end of graph replay
event: some event that is in the middle of a recorded graph

?

I think that would make sense, but I'm not sure if this was explicitly noted in any of the places, unless I overlooked something (which might have happened, ofc).

Reply:
EDIT: I've checked the code sample and I don't think I understand this

```
auto nodeB = graphB.add([&](handler& CGH){
  CGH.depends_on(dynEvent);
```

here, we are depending on event dependency from another recorded graph; is this ok because the event is wrapped into "dynamic" event or, maybe, because the first graph was an explicit graph, with node added via `graphA.add(...);` ?

Reply:
I agree the wording around these types of events is a little confusing and I will try to push some changes soon to clarify, but perhaps I can clarify here as well.

>  Is this a difference between:
>
> event: end of graph replay
> event: some event that is in the middle of a recorded graph

Yes there is a difference here. Events returned from a recorded queue submission to a graph, as in:

```c++
Graph.begin_recording(Queue);
auto EventA = Queue.submit(...);

Queue.submit((sycl::handler& CGH){
// Valid because it came from a previous submission to the same graph
CGH.depends_on(EventA);
...
});
Graph.end_recording();
```
These events have no associated device event (because nothing is created on the device until `finalize()` is called) and so we only use them for establishing internal edges within a graph. So they can be used with `handler::depends_on()` inside another submission to the same graph but not waited on or used as dependencies for normal SYCL submissions or a separate graph's nodes/recorded submissions.

Events returned from a submission of a graph, i.e. `auto ExecEvent = queue::ext_oneapi_graph(execGraph);` are normal SYCL events like those returned from other normal queue submissions, and that event represents the completion of the entire graph.

The second example involving dynamic events is a slightly different case. When a graph in the modifiable state is finalized into an executable graph, it is now possible to obtain SYCL events backed by device events using the `command_graph<state::executable>::get_event(Node)` API. This is because `finalize()` is the step that actually creates the device objects for execution, and so it is now possible to look up the device event associated with a particular node in an executable graph, and return it as a SYCL event to the user.  This resulting SYCL event can be used as any normal SYCL event can.

Because it is a normal SYCL event it can be used as a dependency for nodes in another graph (as seen in the example), this creates an "external" event dependency not an internal edge in the graph. External event dependencies can also be created from any other normal SYCL events (such as those returned from normal queue submissions).

To complicate this slightly I am going to be making some changes to the wording around the use of `get_event()` described here based on discussion [here](https://github.com/reble/llvm/pull/372#discussion_r1700591179) which may change slightly how this works for establishing dependencies between different graphs. Allowing the use of the events I first described (returned from recorded queue submissions, with no device event backing) to establish dependencies between graphs directly has also been discussed [here](https://github.com/reble/llvm/pull/372#discussion_r1617262562), but that would likely be done as a future update to this part of the spec not as part of this initial work.

Sorry for the very verbose reply but hopefully that clarifies things as they stand right now! Please let me know if you have further comments or questions though.